### PR TITLE
sz: new version 2.0.2.0

### PR DESCRIPTION
**Pull request by disheng222**
In this version, we have important improvements on lossy compression ratio especially for high-compression cases. Details can be found in our github site. 
The previous version is 1.4.13.5.

### CODE REVIEW

File: var/spack/repos/builtin/packages/sz/package.py [hunk 1]
```diff
@@ -30,10 +30,11 @@ class Sz(AutotoolsPackage):
     """Error-bounded Lossy Compressor for HPC Data."""
 
     homepage = "https://collab.cels.anl.gov/display/ESR/SZ"
-    url      = "https://github.com/disheng222/SZ/archive/v1.4.13.5.tar.gz"
+    url      = "https://github.com/disheng222/SZ/archive/v2.0.2.0.tar.gz"
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/sz/package.py]: Just a minor remark: in general an update of the `url` is not needed if the "structure" of the `url` didn't change.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/sz/package.py]: I always update the URL, because OCD. But yes, it isn't necessary.

**Comment by adamjstewart** [comment]
Can you rebase on the latest tip of develop now that #3161 has been merged? It modified every package that downloads from Git.

**Comment by disheng222** [comment]
Hi,
I don't quite understand what I should do.
Please be more specific. I'll do whatever I need to do asap.
Thanks!

Best,
Sheng

**Comment by adamjstewart** [comment]
You either need to merge Spack's current develop branch into your PR branch, or rebase your branch on top of Spack's current develop branch. If you're new to Git, this is the tutorial I used to learn it:
https://www.atlassian.com/git/tutorials/merging-vs-rebasing

**Comment by adamjstewart** [comment]
I always forget that I can resolve merge conflicts myself. Should be good to go now as soon as the tests pass.

**Comment by disheng222** [comment]
Hi,
I am confused about the generated file package.py.
It seems that the content is changes a lot from the previously generated
ones.
For example,
Previous ones look like this:
    version('2.0.2.0', '3d63584e2feae5ae6dd21363761cfc0b')
    version('1.4.13.5', 'a2f6147c3c74d74c938dd17d914a4cb8')
    version('1.4.13.4', '1c47170a9eebeadbf0f7e9b675d68d76')
    version('1.4.12.3', '5f51be8530cdfa5280febb410ac6dd94')
    version('1.4.11.0', '10dee28b3503821579ce35a50e352cc6')
    version('1.4.10.0', '82e23dc5a51bcce1f70ba7e3b68a5965')
    version('1.4.9.2',  '028ce90165b7a4c4051d4c0189f193c0')
But the new one look as follows:
    version('2.0.2.0',
sha256='59d1b51d4461049a67e6390a259549eba37f53e20c2d97e3189f22a46d8e6fd4')
    version('1.4.13.5',
sha256='b5e37bf3c377833eed0a7ca0471333c96cd2a82863abfc73893561aaba5f18b9')
    version('1.4.13.4',
sha256='c99b95793c48469cac60e6cf82f921babf732ca8c50545a719e794886289432b')
    version('1.4.13.3',
sha256='9d80390f09816bf01b7a817e07339030d596026b00179275616af55ed3c1af98')
    version('1.4.13.2',
sha256='bc45329bf54876ed0f721998940855dbd5fda54379ef35dad8463325488ea4c6')
    version('1.4.13.0',
sha256='baaa7fa740a47e152c319b8d7b9a69fe96b4fea5360621cdc96cb250635f946f')


Is it correct?
Thanks.
Best,
Sheng

**Comment by adamjstewart** [comment]
The new version is also correct, just make sure to add back the `develop` version and fix the flake8 errors that were introduced.

This change was made in #8777. Spack previously used MD5 for verifying the integrity of a download, but MD5 can be broken, so we've started to make the switch to SHA256 as a more secure hashing method. Both currently work, but at some point we plan to switch all existing packages to SHA256.

**Comment by disheng222** [comment]
Hi, I am a little confused. I pulled the new version of package.py from spack, but it seems that it still points to the old version 1.4.13.5. 
I finally figured out how to sync my forked repo with the spack source repo. I already prepared a new pull request, but not sure if I should submit a new request, or you will deal with the merge conflict? 
Thanks.

Sheng

**Comment by adamjstewart** [comment]
That might be my fault. I already resolved the merge conflict, which automatically made another commit to your origin, but you probably committed to your local branch without that merge commit.

Whatever is easiest for you. You can force push over your current commit or close this PR and create a new one.

File: var/spack/repos/builtin/packages/sz/package.py [hunk 2]
```diff
@@ -51,3 +58,4 @@ def configure_args(self):
         else:
             args += ['--disable-fortran']
         return args
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sz/package.py]: Pretty sure this extra newline will cause flake8 to crash.

**Comment by disheng222** [comment]
Hi,
The first draft of package.py for SZ in Spack was edited by someone else. I
guess he/she was in the Spack team or so. This part related to
--disable-fortran was added by a him/her. I never changed this part since
then.
We can remove this fortran part if that may lead to errors.

Best,
Sheng

**Comment by adamjstewart** [comment]
I think it's fine to keep the fortran part unless you're experiencing problems with it. You just need to remove the empty line at the end of the file to get flake8 to pass.

**Comment by ax3l** [comment]
@disheng222 I think this PR broke the package.
Building `spack install sz` fails on Ubuntu 18.04 with GCC 8.3.0 with:
```
==> Error: ProcessError: Command exited with status 2:
    'make' '-j8'

3 errors found in build log:
     341    ar: `u' modifier ignored since `D' is the default (see `U')
     342    libtool: link: ranlib .libs/libzstd.a
     343    libtool: link: ( cd ".libs" && rm -f "libzstd.la" && ln -s "../libzstd.la" "libzstd.la" 
            )
     344    make[2]: Leaving directory '/tmp/axel/spack-stage/spack-stage-sz-2.0.2.0-xu7c4aylaanrxkv
            cg26awdsrpcmf55tm/spack-src/zstd'
     345    Making all in sz
     346    make[2]: Entering directory '/tmp/axel/spack-stage/spack-stage-sz-2.0.2.0-xu7c4aylaanrxk
            vcg26awdsrpcmf55tm/spack-src/sz'
  >> 347    make[2]: *** No rule to make target 'sz.mod', needed by 'all-am'.  Stop.
     348    make[2]: *** Waiting for unfinished jobs....
     349    make[2]: Leaving directory '/tmp/axel/spack-stage/spack-stage-sz-2.0.2.0-xu7c4aylaanrxkv
            cg26awdsrpcmf55tm/spack-src/sz'
     350    Makefile:404: recipe for target 'all-recursive' failed
  >> 351    make[1]: *** [all-recursive] Error 1
     352    make[1]: Leaving directory '/tmp/axel/spack-stage/spack-stage-sz-2.0.2.0-xu7c4aylaanrxkv
            cg26awdsrpcmf55tm/spack-src'
     353    Makefile:336: recipe for target 'all' failed
  >> 354    make: *** [all] Error 2
```

**Comment by ax3l** [comment]
@disheng222 do you like to add yourself in the package after the `url=` line via `maintainers = ['disheng222']`? :) This would allow pull-request reviewers to ping you more often when changes to the `sz` package are proposed.

**Comment by disheng222** [comment]
Hi Axel,
I tested SZ package and found the following phenomenon. In fact, if you enable **Fortran** (such as ./configure --enable-fortran) and then use make "-j", then the compilation would fail like that; but simply executing 'make -j' again can finish the test compilation successfully. Alternatively, if you enable Fortran option and use "make" (without -j), then the compilation can always succeed once for all. It's unclear to me what happened. But, could you set your configuration to be like: when enabling Fortran, compile SZ using single core (i.e., without -j)? That may take longer but avoid the potential issue. Based on my testing, if compiling without Fortran option, then 'make -j' seems always no problems. 

I will add maintainers = ['disheng222'].

-
Best,
Sheng

**Comment by ax3l** [comment]
Hi Sheng,

thank you for the details!
That sounds very much like a missing dependency in the sz build system that is causing a race condition on compilation targets.

While we can work-around this in Spack by disabling parallel compiles, it's probably something to try to fix upstream. I would love to help you with that but my build system knowledge is very CMake-centric and I have no real idea about autotools.

In the meantime: yes, when you add a PR adding `maintainers = ['disheng222']` just add `parallel = False` as well for now to the package, which disables parallel builds.

**Pull request merged**